src/Pate/AssumptionSet.hs

```diff
@@ -0,0 +1,340 @@
+  apply' sym cache asm f
+
+-- | Rewrite the given 'f' with any bindings in the given 'AssumptionSet'.
+--   Bindings are applied repeatedly to each component expression until a fixpoint
+--   is reached or a loop is detected.
+apply' ::
+  forall sym f m t solver fs.
+  PEM.ExprMappable sym f =>
+  IO.MonadIO m =>
+  sym ~ (W4B.ExprBuilder t solver fs) =>
+  sym ->
+  VarBindCache sym ->
+  AssumptionSet sym ->
+  f ->
+  m f
+apply' sym cache asm f = do
+  let
+    doRebind :: forall tp. ExprSet sym tp -> W4.SymExpr sym tp -> m (W4.SymExpr sym tp)
+    doRebind ancestors e = do
+      e' <- rebindWithFrame' sym cache asm e
+      case SetF.member e' ancestors of
+        True -> return e'
+        False -> doRebind (SetF.insert e' ancestors) e'
+  PEM.mapExpr sym (doRebind mempty) f
+
+-- | Augment an assumption set by first rewriting its entries with the given
+--   assumptions before adding them to the set.
+--   i.e. given `a + b -> {c,d}` and adding `a -> b; c -> d` the result
+--   will be `b + b -> {d}; a -> b; c -> d`
+augment ::
+  sym ~ (W4B.ExprBuilder t solver fs) =>
+  IO.MonadIO m =>
+  sym ->
+  AssumptionSet sym {- ^ original asm set -}  ->
+  AssumptionSet sym {- ^ additional assumptions -} ->
+  m (AssumptionSet sym)
+augment sym origAsm newAsm = do
+  cache <- IO.liftIO WEH.freshVarBindCache
+  origAsm' <- PEM.mapExpr sym (apply' sym cache newAsm) origAsm
+  return $ newAsm <> origAsm'
+
+-- | Retrieve a value that the given expression is bound to in
+--   the given 'AssumptionSet'.
+--   In the case of multiple results, constant values are preferred,
+--   otherwise the least element of all possible symbolic bindings is returned.
+getSomeBinding ::
+  forall sym tp.
+  W4.IsSymExprBuilder sym =>
+  OrdF (W4.SymExpr sym) =>
+  sym ->
+  AssumptionSet sym ->
+  W4.SymExpr sym tp ->
+  Maybe (W4.SymExpr sym tp)
+getSomeBinding sym asm e =
+  let allBinds = getAllBindings sym SetF.empty asm e
+  in case SetF.lookupMin (SetF.filter (\e' -> isJust (W4.asConcrete e')) allBinds) of
+    Just e' -> Just e'
+    Nothing -> SetF.lookupMin allBinds
+
+-- | Retrieve all possible bindings (transitively) for a given expression.
+getAllBindings ::
+  forall sym tp.
+  W4.IsSymExprBuilder sym =>
+  sym ->
+  SetF (W4.SymExpr sym) tp ->
+  AssumptionSet sym ->
+  W4.SymExpr sym tp ->
+  SetF (W4.SymExpr sym) tp
+getAllBindings sym ancestors asm e = case SetF.member e ancestors of
+  True -> SetF.empty
+  False -> case MapF.lookup e (asmBinds asm) of
+    Just es ->
+      let ancestors' = SetF.insert e ancestors
+      in SetF.unions (map (\e' -> SetF.insert e' (getAllBindings sym ancestors' asm e')) (SetF.toList es))
+    Nothing -> case W4.exprType e of
+      W4.BaseBoolRepr | isAssumedPred asm e -> SetF.singleton (W4.truePred sym)
+      _ -> SetF.empty
+
+-- | Compute a predicate that collects the individual assumptions in the frame, including
+-- equality on all bindings.
+toPred ::
+  forall sym m.
+  W4.IsSymExprBuilder sym =>
```

Comment:
Can you state precisely when the returned predicate is true? Is it true if all of the assumptions hold if you assume the initial equalities on bindings?